File: ocaml/2015/day03b/day03b.ml
open Printf
open Sys

module Position2D =
  struct
    type t = int * int
    let compare (x0, y0) (x1, y1) =
      match Stdlib.compare x0 x1 with
      | 0 -> Stdlib.compare y0 y1
      | res -> res
  end

module PositionSet = Set.Make(Position2D)

let usage progname =
  eprintf "usage: %s <input file>\n" progname;
  exit 1

let process_file filename =
  let process_ch (santa, robo_santa, positions, santa_move) ch = (
    let (santa_x, santa_y) = santa in
    let (robo_santa_x, robo_santa_y) = robo_santa in
    let new_santa = if santa_move then
                      ( match ch with
                        | '^' -> (santa_x, santa_y + 1)
                        | 'v' -> (santa_x, santa_y - 1)
                        | '<' -> (santa_x - 1, santa_y)
                        | '>' -> (santa_x + 1, santa_y)
                        | _ -> santa )
                    else santa in
    let new_robo_santa = if santa_move then
                           robo_santa
                         else
                           ( match ch with
                             | '^' -> (robo_santa_x, robo_santa_y + 1)
                             | 'v' -> (robo_santa_x, robo_santa_y - 1)
                             | '<' -> (robo_santa_x - 1, robo_santa_y)
                             | '>' -> (robo_santa_x + 1, robo_santa_y)
                             | _ -> robo_santa ) in
    let new_positions = if santa_move then
                          (PositionSet.add new_santa positions)
                        else
                          (PositionSet.add new_robo_santa positions) in
    (new_santa, new_robo_santa, new_positions, not santa_move)
  )
  in
  try
    let content = In_channel.input_all (In_channel.open_text filename) in
    let (_, _, final_positions, _) = String.fold_left process_ch ((0, 0), (0, 0), PositionSet.empty, true) content in
    PositionSet.cardinal final_positions
  with
  | Sys_error msg ->
      eprintf "error reading file: %s\n" msg;
      exit 1

let () =
  let args = Sys.argv
  in if Array.length args < 2 then
       usage args.(0) ;
     let filename = args.(1) in
     let result = process_file filename in
     printf "result = %d\n" result

File: nim/2015/day03b/day03b.nim
import os, sets

type
  Position2D = object
    x, y: int

proc `==`(a, b: Position2D): bool =
  a.x == b.x and a.y == b.y

proc hash(a: Position2D): int =
  (a.x shl 16) or a.y

proc processFile(filename: string): int =
  var positions = initHashSet[Position2D]()
  var santa = Position2D(x: 0, y: 0)
  var roboSanta = Position2D(x: 0, y: 0)
  var santaMove = true
  positions.incl(santa)
  for line in lines(filename):
    for ch in line:
      if santaMove:
        if ch == '^': santa = Position2D(x: santa.x, y: santa.y + 1)
        elif ch == 'v': santa = Position2D(x: santa.x, y: santa.y - 1)
        elif ch == '<': santa = Position2D(x: santa.x - 1, y: santa.y)
        elif ch == '>': santa = Position2D(x: santa.x + 1, y: santa.y)
        positions.incl(santa)
      else:
        if ch == '^': roboSanta = Position2D(x: roboSanta.x, y: roboSanta.y + 1)
        elif ch == 'v': roboSanta = Position2D(x: roboSanta.x, y: roboSanta.y - 1)
        elif ch == '<': roboSanta = Position2D(x: roboSanta.x - 1, y: roboSanta.y)
        elif ch == '>': roboSanta = Position2D(x: roboSanta.x + 1, y: roboSanta.y)
        positions.incl(roboSanta)
      santaMove = not santaMove
  return positions.len()

when isMainModule:
  if paramCount() < 1:
    quit "usage: " & getAppFilename() & " <input file>", 1

  let filename = paramStr(1)
  try:
    echo "result = ", processFile(filename)
  except OSError as e:
    quit "error while processing file `" & filename & "': " & e.msg, 1
